### DISCUSSION TITLE
Modify base_url or _email_link generation on a per action basis

### DISCUSSION THREAD

**Discussion by FelipeBodelon** [janko/rodauth-rails#278] [Q&A] [answered]
I'm handling some callback routes or email links on different hosts depending on the action, as I'm going for a progressive implementation of the library. I want to keep the generated rails view, while creating my custom auth flow on a new host vía JSON API. Is there a way to modify `base_url` on a per action basis? So far I mostly need the verify email to have a different `base_url`. 

I've tried this and a few other overrides on `rodauth_mailer.rb` but they don't seem to be changing the default:
```rb
# rodauth_main.rb

def verify_account_email_link
    "#{ENV.fetch("FRONTEND_APP_URL", "http://localhost:3001")}/auth/verify-account?#{verify_account_key_param}=#{verify_account_key_value}"
end
```

**Comment by janko** [comment]
That method definition should be overriding the email link. In your mailer, when you look at `@rodauth.method(:verify_account_email_link).source_location`, it doesn't show you overridden method? Where is your method definition located in your Rodauth configuration, outside of the `configure {}` block or within `auth_class_eval {}`?

**Reply by FelipeBodelon** [reply]
It is located inside the `configure {}` block. 
I get `["/usr/share/rvm/gems/ruby-3.2.2/gems/rodauth-2.33.0/lib/rodauth/features/verify_account.rb", 204]` for the `.source_location`

**Reply by FelipeBodelon** [reply]
Right before I tried modifying `base_url` with success. Maybe I can conditionally change it there if I have access to the action in that method (since is the only thing I need to change):

```rb
    # base_url do
    #   modify conditionally here
    #   ENV.fetch("FRONTEND_APP_URL", "http://localhost:3001")
    # end

    def verify_account_email_link
      "#{ENV.fetch("FRONTEND_APP_URL", "http://localhost:3001")}/auth/verify-account?#{verify_account_key_param}=#{verify_account_key_value}"
    end
```

**Reply by janko** [reply]
If it's located directly in the `configure {}` block, then it won't work. In the `configure {}` block you're inside Rodauth configuration DSL, not in the context of the auth class. You need to either place it outside:

```rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    # ...
  end

  def verify_account_email_link
    # ...
  end
end
```

Or just use the configuration DSL:

```rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    verify_account_email_link do
      # implementation
    end
  end
end
```

**Reply by FelipeBodelon** [reply]
I was using the configuration DSL, sorry if I didn't said it clearly. Tried placing it outside and the override worked (not with the DSL tho). I found another resulting issue in regards to the actual verification on the callback route (external host, via JSON). Verification fails with a 401 error. My intuition suggests that it could fail as the origin of the request wouldn't match `base_url`, which further nudges me into believing overriding `base_url` conditionally could better solve my use case.

Is this a reason why it could be failing? If I override `base_url` then it doesn't fail.

**Reply by janko** [reply]
>  Tried placing it outside and the override worked (not with the DSL tho).

Note sure why it didn't work via the configuration DSL, works on the official demo app 🤷🏻‍♂️ 

> Verification fails with a 401 error.

What is the response body? Is it coming from Rodauth or Rails' CSRF protection? With the current information, I don't know why overriding `base_url` makes it not fail.

**Reply by FelipeBodelon** [reply]
Rodauth CSRF is disabled. Fails with status 401 `{ error: "Unable to verify account" }`. Seems to be coming from Rodauth, but I have no idea what could be the issue.

**Reply by janko** [reply, marked as answer]
Since there are no validation errors, this most likely means the token was invalid. You can verify that by setting error reason in the JSON response, then checking if it matches `invalid_verify_account_key`.

```rb
set_error_reason { |reason| json_response[:error_reason] = reason }
```

The reason it's invalid is because rodauth-rails configures Rodauth to use [HMACs](https://github.com/jeremyevans/rodauth?tab=readme-ov-file#label-HMAC), which requires hashed tokens to be passed to email links, and you're passing the raw token. That's why overriding `base_url` was working for you, because it wasn't changing the token. So, it's have to be

```rb
def verify_account_email_link
  "#{ENV.fetch("MEMOORIAS_APP_URL", "http://localhost:3001")}/auth/verify-account?#{verify_account_key_param}=#{token_param_value(verify_account_key_value)}"
end
```

But I would recommend just overriding the `base_url` based on a custom flag that you turn on in your mailer actions:

```rb
class RodauthMailer < ApplicationMailer
  def verify_account(name, account_id, key)
    @rodauth = rodauth(name, account_id) { @verify_account_key_value = key }
    @rodauth.frontend = true
    @account = @rodauth.rails_account

    mail subject: @rodauth.verify_account_email_subject
  end
end
```
```rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    base_url { frontend ? ENV.fetch("MEMOORIAS_APP_URL", "http://localhost:3001") : super() }
  end

  attr_accessor :frontend
end
```

**Reply by FelipeBodelon** [reply]
That's quite exactly the solution I needed. Thank you very much! This gives me full control based on action and modifies only the necessary behaviours without altering the rest!
